File: examples/gmail-inbox-node/inbox.mjs
import {google} from "googleapis";

const USAGE = `gmail-inbox: run this app under Fabricate.

  fab run acme-gmail -- node inbox.mjs
  fab run acme-gmail --proxy -- node inbox.mjs

Optional search: node inbox.mjs INV-4812
`;

function env(...names) {
  for (const name of names) {
    const value = process.env[name];
    if (value) return value;
  }
  return undefined;
}

function usingProxy() {
  return Boolean(env("HTTPS_PROXY", "https_proxy"));
}

function gmailClient() {
  const auth = new google.auth.OAuth2("fab-client", "fab-secret");
  if (usingProxy()) {
    // Official client keeps oauth2.googleapis.com and gmail.googleapis.com.
    // Fabricate answers the refresh grant and the Gmail calls locally.
    auth.setCredentials({
      access_token: "expired",
      expiry_date: Date.now() - 60_000,
      refresh_token: "fab-refresh-token",
    });
    return google.gmail({version: "v1", auth});
  }
  const token = env("FAB_SUPPORT_MAIL_TOKEN", "FAB_GMAIL_TOKEN");
  const serviceURL = env("FAB_SUPPORT_MAIL_URL", "FAB_GMAIL_URL");
  if (!token || !serviceURL) {
    console.error(USAGE);
    process.exit(1);
  }
  auth.setCredentials({
    access_token: token,
    expiry_date: Date.now() + 60 * 60 * 1_000,
  });
  return google.gmail({
    version: "v1",
    auth,
    rootUrl: `${serviceURL.replace(/\/$/, "")}/`,
  });
}

function header(message, name) {
  const headers = message.data?.payload?.headers ?? [];
  const match = headers.find((item) => item.name?.toLowerCase() === name.toLowerCase());
  return match?.value ?? "";
}

const query = process.argv.slice(2).join(" ").trim();
const gmail = gmailClient();
const profile = await gmail.users.getProfile({userId: "me"});
const listed = await gmail.users.messages.list({
  userId: "me",
  maxResults: 8,
  ...(query ? {q: query} : {}),
});

const email = profile.data.emailAddress ?? "unknown";
const total = profile.data.messagesTotal ?? "?";
const estimate = listed.data.resultSizeEstimate ?? 0;
const messages = listed.data.messages ?? [];
let title = `${email}  ·  ${total} messages`;
if (query) title += `  ·  search ${JSON.stringify(query)} (${estimate})`;
console.log(title);
console.log();
if (messages.length === 0) {
  console.log("  (no messages)");
  process.exit(0);
}
for (const item of messages) {
  const message = await gmail.users.messages.get({userId: "me", id: item.id});
  const subject = header(message, "Subject") || "(no subject)";
  const sender = header(message, "From") || "(unknown sender)";
  const snippet = (message.data.snippet ?? "").trim();
  console.log(`  ${subject}`);
  console.log(`    ${sender}`);
  if (snippet) console.log(`    ${snippet}`);
  console.log();
}

File: examples/gmail-inbox-python/inbox.py
#!/usr/bin/env python3
"""List the Acme Gmail support inbox. Run under `fab run acme-gmail`."""

from __future__ import annotations

import json
import os
import ssl
import sys
import urllib.error
import urllib.parse
import urllib.request

USAGE = """\
gmail-inbox: run this app under Fabricate.

  fab run acme-gmail -- python3 inbox.py
  fab run acme-gmail --proxy -- python3 inbox.py

Optional search: python3 inbox.py INV-4812
"""


def env(*names: str) -> str | None:
    for name in names:
        value = os.environ.get(name)
        if value:
            return value
    return None


def using_proxy() -> bool:
    return bool(env("HTTPS_PROXY", "https_proxy"))


def gmail_base() -> tuple[str, str | None]:
    if using_proxy():
        return "https://gmail.googleapis.com/gmail/v1", None
    url = env("FAB_SUPPORT_MAIL_URL", "FAB_GMAIL_URL")
    token = env("FAB_SUPPORT_MAIL_TOKEN", "FAB_GMAIL_TOKEN")
    if not url or not token:
        raise SystemExit(USAGE)
    return f"{url.rstrip('/')}/gmail/v1", token


def ssl_context(url: str) -> ssl.SSLContext | None:
    if not url.startswith("https://"):
        return None
    ctx = ssl.create_default_context()
    ca = env("SSL_CERT_FILE", "REQUESTS_CA_BUNDLE")
    if ca:
        ctx.load_verify_locations(ca)
    return ctx


def get_json(url: str, token: str | None) -> object:
    headers = {"Accept": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    request = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(request, context=ssl_context(url), timeout=30) as response:
            return json.load(response)
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace")
        raise SystemExit(f"gmail-inbox: {exc.code} {url}\n{detail}") from exc


def header(payload: dict, name: str) -> str:
    for item in payload.get("payload", {}).get("headers", []) or []:
        if str(item.get("name", "")).lower() == name.lower():
            return str(item.get("value") or "")
    return ""


def main() -> None:
    base, token = gmail_base()
    query = " ".join(sys.argv[1:]).strip()
    profile = get_json(f"{base}/users/me/profile", token)
    params = {"userId": "me", "maxResults": "8"}
    if query:
        params["q"] = query
    listed = get_json(f"{base}/users/me/messages?{urllib.parse.urlencode(params)}", token)

    email = profile.get("emailAddress", "unknown")
    total = profile.get("messagesTotal", "?")
    estimate = listed.get("resultSizeEstimate", 0)
    messages = listed.get("messages") or []
    title = f"{email}  ·  {total} messages"
    if query:
        title += f"  ·  search {query!r} ({estimate})"
    print(title)
    print()
    if not messages:
        print("  (no messages)")
        return
    for item in messages:
        message = get_json(f"{base}/users/me/messages/{item['id']}", token)
        subject = header(message, "Subject") or "(no subject)"
        sender = header(message, "From") or "(unknown sender)"
        snippet = (message.get("snippet") or "").strip()
        print(f"  {subject}")
        print(f"    {sender}")
        if snippet:
            print(f"    {snippet}")
        print()


if __name__ == "__main__":
    main()
